### DISCUSSION TITLE
Toggle IsSensor after body is added to physics system

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1003] [Q&A] [answered]
I hope I am not being annoying and find it at least slightly entertaining the ways I'm trying to break Jolt. :)

When I set body.SetIsSensor(true) after it has already been added to the physics system (and some simulations have potentially already run), I get an assert failure at:
https://github.com/jrouwe/JoltPhysics/blob/0191c72d5f6a66217df02c0f6ad5d1c63134b4d4/Jolt/Physics/Collision/CollisionDispatch.cpp#L23

This is the code that causes the assert failure:
`
	JPH::BodyLockWrite lock(mPhysicsSystem->GetBodyLockInterface(), object->collisionObjectID);
	if (lock.Succeeded())
	{
		JPH::Body &body = lock.GetBody();
		body.SetIsSensor(isSensor);
	}
`

Is this simply not supported? What would be the solution, to remove the body from the scene and re-insert it? I tried doing so (please try to not bang your desk against the desk at my crazy code), but it still results in the same assert error:

`
void CollisionHandler::SetAsSensor(WObj::WObj *object, bool isSensor)
{
	if (object->collisionObjectID.IsInvalid())
		return;

	bool doSwap = false;
	{
		JPH::BodyLockWrite lock(mPhysicsSystem->GetBodyLockInterface(), object->collisionObjectID);
		if (lock.Succeeded())
		{
			JPH::Body &body = lock.GetBody();

			if (body.IsSensor() != isSensor)
				doSwap = true;
		}
	}

        // Gotta scope all of these locks properly so we don't get double-locking...
	if (doSwap)
	{
		JPH::BodyInterface &bi = mPhysicsSystem->GetBodyInterface();
		bi.RemoveBody(object->collisionObjectID);

		{
			JPH::BodyLockWrite lock(mPhysicsSystem->GetBodyLockInterface(), object->collisionObjectID);
			if (lock.Succeeded())
			{
				JPH::Body &body = lock.GetBody();

				body.SetIsSensor(isSensor);
			}
		}

		bi.AddBody(object->collisionObjectID, JPH::EActivation::Activate);
	}
}
`

I am trying to change the IsSensor flag on a kinematic triangle mesh, btw. This appears to work fine when I use a convex hull instead. Bug or intentional?

**Comment by jrouwe** [comment, marked as answer]
Your error means that you are now trying to (probably) collide a mesh with a mesh which is not supported. Make sure that your mesh sensor is not in a layer that collides with other meshes.

**Reply by SSNTails** [reply]
D'oh, you are absolutely right, I should be creating my moving triangle meshes on a layer separate from the dynamic ones. 🤦‍♂️ While it can be frustrating sometimes, I appreciate all of the asserts you've added to the library to make sure people are doing things right.
